### PR TITLE
omnibus-2024-04-15

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Synopsis**

Four sets of changes:

1. Implement the stdin guard described in [UW-514](https://jira.epic.oarcloud.noaa.gov/browse/UW-514). This is a change to the API routines such that an optional argument, `stdin_ok` must be set when an API user wants to read from stdin. The argument is always set when called from the CLI, so there is no change for CLI users. For API users, this is meant to help avoid accidental read-from-stdin configurations that will lead to `uw` blocking waiting for input, which is especially bad in HPC batch jobs, where walltime would be run out uselessly. With this change, an API call that fails to provide a non-`None` value for the primary data-source variable and that also does not set `stdin_ok` would raise and exception. See below for an example.
2. Also in the API, where `Path` objects are taken as argument values, also accept `str` values, and convert those as needed to `Path` objects. This preserves the `Path`-only status of the rest of the `uwtools` package, but allows API users to supply paths as `str` values, as they may be unaccustomed to working with `pathlib` and surprised by a requirement for `Path` objects.
3. Replace an `assert` statement ensuring that the value parsed from a YAML data source is a `dict` with logic that performs this check and provides gentler, more informative advice to the user. See below for an example.
4. Add a generic `existing()` utility task that accepts files, directories, or symlinks; and make the `symlink()` task depend on it. This allows the `file` tool to symlink directories and other symlinks as well as regular files.

There are a lot of changed lines, but most are repetitions of changes related to items 1 and 2.

Examples:

Item 1: When `stdin_on` is not set via an API call when `config` is also not set, implying a read from stdin:

```
% python -c 'import datetime as dt; from uwtools.api.ungrib import execute; execute(task="run", cycle=dt.datetime.now())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/maddenp/git/uwtools/src/uwtools/api/ungrib.py", line 39, in execute
    config=_ensure_data_source(config, stdin_ok), cycle=cycle, batch=batch, dry_run=dry_run
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/utils/api.py", line 24, in ensure_data_source
    raise UWError("Set stdin_ok=True to permit read from stdin")
uwtools.exceptions.UWError: Set stdin_ok=True to permit read from stdin
```

With `stdin_ok` set:
```
% python -c 'import datetime as dt; from uwtools.api.ungrib import execute; execute(task="run", cycle=dt.datetime.now(), stdin_ok=True)'

<waiting for input on stdin>
```

Item 3: Old behavior when value parsed from YAML is not a `dict`:

```
% uw file copy --target-dir ./target --config-file config.yaml 
Traceback (most recent call last):
  File "/home/maddenp/sw/conda/envs/DEV-uwtools/bin/uw", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/cli.py", line 73, in main
    sys.exit(0 if modes[args[STR.mode]](args) else 1)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/cli.py", line 349, in _dispatch_file
    return actions[args[STR.action]](args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/cli.py", line 358, in _dispatch_file_copy
    return uwtools.api.file.copy(
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/api/file.py", line 27, in copy
    _FileCopier(target_dir=target_dir, config=config, keys=keys, dry_run=dry_run).go()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/file.py", line 40, in __init__
    self._config = YAMLConfig(config=config)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/config/formats/base.py", line 38, in __init__
    self.data = self._load(self._config_file)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maddenp/git/uwtools/src/uwtools/config/formats/yaml.py", line 71, in _load
    assert isinstance(config, dict)
AssertionError
```

New behavior:

```
% uw file copy --target-dir ./target --config-file config.yaml 
[2024-04-12T22:30:22]    ERROR Parsed a str value from config.yaml, expected a dict
```

**Type**

- [x] Bug fix (corrects a known issue)
- [x] Enhancement (adds a new functionality)

**Impact**

- [x] This is a breaking change (changes existing functionality)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: src/uwtools/api/chgres_cube.py [hunk 1]
```diff
@@ -4,21 +4,23 @@
...
 import datetime as dt
 from pathlib import Path
-from typing import Dict, Optional
+from typing import Dict, Optional, Union
 
 import iotaa as _iotaa
 
 import uwtools.drivers.support as _support
 from uwtools.drivers.chgres_cube import ChgresCube as _ChgresCube
+from uwtools.utils.api import ensure_data_source as _ensure_data_source
```
Comment (maddenp-noaa) [inline src/uwtools/api/chgres_cube.py:13]: The `ensure_data_source()` function 1. Enforces the requirement that a `None` data-source value in the API requires that `stdin_ok` is set, and 2. Converts `str` values to `Path` values to maintain the downstream invariant that `uwtools` uses `Path` objects internally.

File: src/uwtools/api/chgres_cube.py [hunk 2]
```diff
@@ -32,9 +34,12 @@ def execute(
     :param batch: Submit run to the batch system?
     :param dry_run: Do not run the executable, just report what would have been done.
     :param graph_file: Write Graphviz DOT output here.
+    :param stdin_ok: OK to read from stdin?
     :return: ``True`` if task completes without raising an exception.
     """
-    obj = _ChgresCube(config=config, cycle=cycle, batch=batch, dry_run=dry_run)
+    obj = _ChgresCube(
+        config=_ensure_data_source(config, stdin_ok), cycle=cycle, batch=batch, dry_run=dry_run
```
Comment (maddenp-noaa) [inline src/uwtools/api/chgres_cube.py:41]: The `config` value will be a `Path` if its was initially a `str`, or otherwise whatever value was originally supplied (a `dict`, `Config`, or `Path`) -- unless it was `None` and `stdin_ok` was not set, in which case an exception will be raised.

A lot of the same follows... I'll comment on interesting changes.

File: src/uwtools/api/config.py [hunk 3]
```diff
@@ -144,25 +173,12 @@ def validate(
...
-    config: Union[dict, _Config, Optional[Path]]
-) -> Union[_Config, Optional[Path]]:
-    """
-    Encapsulate a ``dict`` in a ``Config``; return a ``Config`` or path argument as-is.
-
-    :param config: A config as a ``dict``, ``Config``, or ``Path``
-    """
-    if isinstance(config, dict):
-        return _YAMLConfig(config=config)
-    return config
```
Comment (maddenp-noaa) [inline src/uwtools/api/config.py:165]: This is handled by `ensure_data_source()` now.

File: src/uwtools/cli.py [hunk 4]
```diff
@@ -129,6 +129,7 @@ def _dispatch_chgres_cube(args: Args) -> bool:
         batch=args[STR.batch],
         dry_run=args[STR.dryrun],
         graph_file=args[STR.graphfile],
+        stdin_ok=True,
```
Comment (maddenp-noaa) [inline src/uwtools/cli.py:132]: The CLI sets `stdin_ok` for all API calls it makes.

File: src/uwtools/config/formats/yaml.py [hunk 5]
```diff
@@ -68,8 +69,12 @@ def _load(self, config_file: Optional[Path]) -> dict:
...
             try:
                 config = yaml.load(f.read(), Loader=loader)
-                assert isinstance(config, dict)
-                return config
+                if isinstance(config, dict):
+                    return config
+                raise UWConfigError(
+                    "Parsed a %s value from %s, expected a dict"
+                    % (type(config).__name__, config_file or "stdin")
+                )
```
Comment (maddenp-noaa) [inline src/uwtools/config/formats/yaml.py:77]: As shown in the PR description, this provides an informative failure message for the user instead of a more obscure stacktrace.

File: src/uwtools/tests/api/test_chgres_cube.py [hunk 6]
```diff
@@ -1,26 +1,29 @@
-# pylint: disable=missing-function-docstring,protected-access
+# pylint: disable=missing-function-docstring,protected-access,redefined-outer-name
 
 import datetime as dt
+from pathlib import Path
```
Comment (maddenp-noaa) [inline src/uwtools/tests/api/test_chgres_cube.py:4]: Like the library changes, the test changes are mostly repetitions of a common theme.

File: src/uwtools/tests/config/formats/test_yaml.py [hunk 7]
```diff
@@ -106,20 +107,36 @@ def test_simple(tmp_path):
...
 
 
+def test_constructor_error_not_dict_from_file(tmp_path):
+    # Test that a useful exception is raised if the YAML file input is a non-dict value.
+    tmpfile = tmp_path / "test.yaml"
+    with tmpfile.open("w", encoding="utf-8") as f:
+        f.write("hello")
+    with raises(exceptions.UWConfigError) as e:
+        YAMLConfig(tmpfile)
+    assert f"Parsed a str value from {tmpfile}, expected a dict" in str(e.value)
```
Comment (maddenp-noaa) [inline src/uwtools/tests/config/formats/test_yaml.py:129]: Failure mode when YAML read from a file parses to a non-`dict` value.

File: src/uwtools/tests/config/formats/test_yaml.py [hunk 8]
```diff
@@ -106,20 +107,36 @@ def test_simple(tmp_path):
...
+
+
+def test_constructor_error_not_dict_from_stdin():
+    # Test that a useful exception is raised if the YAML stdin input is a non-dict value.
+    with patch.object(sys, "stdin", new=StringIO("a string")):
+        with raises(exceptions.UWConfigError) as e:
+            YAMLConfig()
+    assert "Parsed a str value from stdin, expected a dict" in str(e.value)
```
Comment (maddenp-noaa) [inline src/uwtools/tests/config/formats/test_yaml.py:137]: Failure mode when YAML read from stdin parses to a non-`dict` value.

File: src/uwtools/utils/api.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+"""
+Support for API modules.
```
Comment (maddenp-noaa) [inline src/uwtools/utils/api.py:2]: A new module providing support routines for API modules. I considered placing these in `uwtools.utils.file` but wasn't sure that was right. If we find these routines are needed by non-API code, we could move them, but this seemed to me good enough for now.

File: src/uwtools/utils/tasks.py [hunk 10]
```diff
@@ -45,6 +56,6 @@ def symlink(target: Path, linkname: Path):
     """
     yield "Link %s -> %s" % (linkname, target)
     yield asset(linkname, linkname.exists)
-    yield file(target)
+    yield existing(target)
```
Comment (maddenp-noaa) [inline src/uwtools/utils/tasks.py:59]: The target of a symlink may now be a file, a directory, or another symlink.

File: src/uwtools/utils/api.py (hunk 9 above)
Comment (WeirAE) [reply src/uwtools/utils/api.py:2]: Hrm - yeah, thats fair, I agree. They feel like `file` utilities, but they're really just used in `api` for now so should be fine

**Comment by benjamin-cash** [comment]
I was able to clone, conda build, and incorporate the PR branch into my test environment, so I feel like I have leveled up my conda skills. :) I was able to confirm that the Path and symlink changes worked as expected.

**Comment by maddenp-noaa** [comment]
@benjamin-cash Excellent, and thank you for testing!

**Comment by benjamin-cash** [comment]
> @benjamin-cash Excellent, and thank you for testing!

I may have run into a corner case where it is still hanging and looking for stdin. I accidentally gave uwtools.api.file.link an empty directory to look in for files to link and it hung. When I keybard interrrupted this was the stack trace:
Traceback (most recent call last):
  File "/work2/02441/bcash/frontera/workflow/ecflow/bin/uw_stage_model_files.py", line 36, in <module>
    uwlink(target_dir=EXP_RENDER['run_dir'], config=EXP_RENDER_FILE,
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/api/file.py", line 56, in link
    _FileLinker(
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/file.py", line 44, in __init__
    self._validate()
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/file.py", line 70, in _validate
    validate_internal(schema_name="files-to-stage", config=self._file_map)
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/config/validator.py", line 35, in validate_internal
    if not validate_yaml(config=config, schema_file=schema_file):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/config/validator.py", line 51, in validate_yaml
    cfgobj = _prep_config(config)
             ^^^^^^^^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/config/validator.py", line 72, in _prep_config
    cfgobj = config if isinstance(config, YAMLConfig) else YAMLConfig(config)
                                                           ^^^^^^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/config/formats/base.py", line 38, in __init__
    self.data = self._load(self._config_file)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/config/formats/yaml.py", line 69, in _load
    with readable(config_file) as f:
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/utils/file.py", line 91, in readable
    yield _stdinproxy()
          ^^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/utils/file.py", line 41, in _stdinproxy
    return StdinProxy()
           ^^^^^^^^^^^^
  File "/work2/02441/bcash/frontera/.conda/envs/uwtools_dev/lib/python3.11/site-packages/uwtools/utils/file.py", line 24, in __init__
    self._stdin = sys.stdin.read()
                  ^^^^^^^^^^^^^^^^
KeyboardInterrupt

**Comment by maddenp-noaa** [comment]
@benjamin-cash Can you please share your config file and the command that produced that stacktrace?

**Comment by benjamin-cash** [comment]
Hi @maddenp-noaa - here you go. It was from an API call, and I've attached a script and config that duplicates the problem

```
import yaml
from uwtools.api.file import copy as uwcopy
from uwtools.api.file import link as uwlink

SUBDIR="exec"
run_dir = "/scratch1/02441/bcash/"
uwcopy(target_dir=run_dir, config="test.yaml", keys=[SUBDIR,"files_to_copy"])
uwlink(target_dir=run_dir, config="test.yaml", keys=[SUBDIR,"files_to_link"])
```

test.yaml
exec:
  files_to_copy:
    bar.txt: bar.txt
  files_to_link:

**Comment by maddenp-noaa** [comment]
@benjamin-cash OK, I can reproduce the issue now, thank you. I just pushed what I think is a fix. Behavior for your case should now be:
```
{} should be non-empty

Failed validating 'minProperties' in schema:
    {'additionalProperties': {'type': 'string'},
     'minProperties': 1,
     'type': 'object'}

On instance:
    {}
Traceback (most recent call last):
  File "/home/maddenp/git/uwtools/a.py", line 7, in <module>
    uwlink(target_dir=run_dir, config="test.yaml", keys=[SUBDIR,"files_to_link"])
  File "/home/maddenp/git/uwtools/src/uwtools/api/file.py", line 56, in link
    _FileLinker(
  File "/home/maddenp/git/uwtools/src/uwtools/file.py", line 44, in __init__
    self._validate()
  File "/home/maddenp/git/uwtools/src/uwtools/file.py", line 70, in _validate
    validate_internal(schema_name="files-to-stage", config=self._file_map)
  File "/home/maddenp/git/uwtools/src/uwtools/config/validator.py", line 36, in validate_internal
    raise UWConfigError("YAML validation errors")
uwtools.exceptions.UWConfigError: YAML validation errors
```
The empty value for the `files_to_link` key is, per the YAML spec, interpreted as a YAML `null`, which is then parsed as a Python `None`, which is then indistinguishable from the `None` default value when `config` is not specified. So, `config` -- which was initially a nice, non-`None` filename -- becomes `None` by following `keys` to the empty value of `files_to_link`. This gets passed schema-validation code which interprets the `None` as "Oh, you want to read stdin!" and blocks waiting for input. The fix just replaces the `None` found at the end of the `keys` chain with an empty `dict` value, which the schema validator then flags as an unacceptable value.

**Comment by maddenp-noaa** [comment]
@benjamin-cash No, I found that other values besides the empty value could also force a read from stdin, so I made a more explicit check and just pushed that. Now, an exception like this one is raised:
```
uwtools.exceptions.UWConfigError: No file map found at exec -> files_to_link
```

**Comment by benjamin-cash** [comment]
> @benjamin-cash No, I found that other values besides the empty value could also force a read from stdin, so I made a more explicit check and just pushed that. Now, an exception like this one is raised:
> 
> ```
> uwtools.exceptions.UWConfigError: No file map found at exec -> files_to_link
> ```

@maddenp-noaa - I installed the new version and can confirm I'm seeing this error now in both the example I sent and my original code. Thanks for the quick fix!

...it's nice to see that my ability to stumble into weird corner cases remains undimmed... ;)

File: src/uwtools/tests/api/test_config.py [hunk 11]
```diff
@@ -89,23 +103,4 @@ def test_validate_config_file(tmp_path):
...
-    assert isinstance(config_obj, YAMLConfig)
-    assert config_obj.data == config_dict
-
-
-def test__ensure_config_arg_type_path():
-    config_path = Path("/path/to/config.yaml")
-    config_obj = config._ensure_config_arg_type(config=config_path)
-    assert isinstance(config_obj, Path)
-    assert config_obj is config_path
+    _validate_yaml.assert_called_once_with(schema_file=Path(kwargs["schema_file"]), config=cfg)
```
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/api/test_config.py:106]: For this assertion what is the difference between this option and `schema_file=Path(kwargs["schema_file"]), config=kwargs["config"]` like the test above. It seems like one is access the dict contents and the other is useing the yaml itself.

Comment (maddenp-noaa) [reply src/uwtools/tests/api/test_config.py:106]: The previous (parametrized) test plus this one comprise three tests: One where the config is a `dict` object, again with a `YAMLConfig` object, and then with an on-disk file. The first two both use in-memory objects and were easy to parametrize together. The third test needs extra code to create the on-disk file before testing against it. I'm not certain that's what you were asking -- please let me know if not.

**Review by NaureenBharwaniNOAA — APPROVED** [review]
Looks great to me!

**Review by WeirAE — APPROVED** [review]
Even edge cases solved and confirmed, looks great to me

**Review by christinaholtNOAA — APPROVED** [review]
LGTM! Thanks for taking on these tedious tasks!
